### PR TITLE
Sketching out a "types and shapes" developer document.

### PR DESCRIPTION
**Pull request by ScottTodd**
Progress on https://github.com/google/iree/issues/5223

This is light on details right now, but gives us a place to fill in more over time.

### CODE REVIEW

File: docs/developers/design_docs/types_and_shapes.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+* TensorFlow: [Introduction to Tensors](https://www.tensorflow.org/guide/tensor)
+* PyTorch: [`torch.Tensor` documentation](https://pytorch.org/docs/stable/tensors.html)
+* NumPy: [Data types documentation](https://numpy.org/doc/stable/user/basics.types.html)
+
+Backend references:
+
+* Vulkan: [buffer and image formats](https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#formats)
+* SPIR-V: [types](https://www.khronos.org/registry/SPIR-V/specs/1.0/SPIRV.html#_types) and [capabilities](https://www.khronos.org/registry/SPIR-V/specs/1.0/SPIRV.html#_a_id_capability_a_capability)
+
+## Types
```
Comment (benvanik) [inline docs/developers/design_docs/types_and_shapes.md:32]: would be worth having a section on shapes that mentions that we need ranked shapes and that we (generally) don't care about the rank but there are definitely benefits to using certain structures (same as those for any efficient compute) - don't do a bunch of 10000x3x9x17x3 kind of stuff, and dynamic dimensions are best as outer ones (?x64x128 is better than 128x64x?) - like how you would want int32_t[512] not std::vector<uint8_t>.

**Review by benvanik — CHANGES_REQUESTED** [review]

File: docs/developers/design_docs/types_and_shapes.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+Types are least constrained in user code within high level frameworks, where
+composite types such as Python classes, media files, Protocol Buffers, JSON
+objects, and other data structures can be freely created and transformed.
+Meanwhile, types are most constrained by hardware and device APIs, where only
+specific low level primitives are defined or where certain operations are
+supported by efficient hardware implementations.
+
+### Conversion process
+
+IREE lowers programs from representations produced by high level frondends down
```
Comment (rsuderman) [inline docs/developers/design_docs/types_and_shapes.md:52]: Its probably worth mentioning that we don't directly ingest Tensorflow and ingest TOSA instead. Our TOSA does not require fully defined shapes and we run shape inference after import. It will fully infer as far as it can, however not all cases on all ops are possible. E.g. transpose conv needs to know its shape at construction.

File: docs/developers/design_docs/types_and_shapes.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+Import dialects (`iree`, `tensor`, `linalg`, etc.)
+
+                                    ↓
+
+`flow` dialect (tensor program modeling and compute workload partitioning)
+
+                                    ↓
+
+               `stream` dialect                    |      code generation
```
[the remark this answers is not in the document]
Comment (ScottTodd) [reply docs/developers/design_docs/types_and_shapes.md]: Whoops :P I shuffled that section around a bit. Could use some work on listing the requirements for device and host code generation within the HAL dialect (and codegen dialects like LLVM / SPIR-V)

File: docs/developers/design_docs/types_and_shapes.md (hunk 1 above)
Comment (ScottTodd) [reply docs/developers/design_docs/types_and_shapes.md:36]: Yep! Started filling in the shapes section and shuffled it around so it's ahead of the conversion flow section. How's this look now?

File: docs/developers/design_docs/types_and_shapes.md (hunk 2 above)
Comment (ScottTodd) [reply docs/developers/design_docs/types_and_shapes.md:116]: Mentioned that a bit in the shapes section, thanks. Haven't yet gone into detail on each item down in this section of the doc.

File: docs/developers/design_docs/types_and_shapes.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+Frontend references:
+
+* TensorFlow: [Introduction to Tensors](https://www.tensorflow.org/guide/tensor)
+* PyTorch: [`torch.Tensor` documentation](https://pytorch.org/docs/stable/tensors.html)
+* NumPy: [Data types documentation](https://numpy.org/doc/stable/user/basics.types.html)
+
+Backend references:
+
+* Vulkan: [buffer and image formats](https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#formats)
+* SPIR-V: [types](https://www.khronos.org/registry/SPIR-V/specs/1.0/SPIRV.html#_types) and [capabilities](https://www.khronos.org/registry/SPIR-V/specs/1.0/SPIRV.html#_a_id_capability_a_capability)
```
Comment (hanhanW) [inline docs/developers/design_docs/types_and_shapes.md:34]: Do we want to split this to codegen backends and drivers?

| Backend | Driver|
|----------|------|
| llvm-aot | dylib |
| llvm-aot | dylib-sync |
| spir-v | vulkan |
| cuda | cuda |

File: docs/developers/design_docs/types_and_shapes.md (hunk 2 above)
Comment (hanhanW) [reply docs/developers/design_docs/types_and_shapes.md:116]: note: +1 on TF side. We will infer the shapes as far as it can after importing the model into MLIR.

This is the pass doing shape inference: https://github.com/google/iree/blob/94a2168c63ac5e075be132fe1b97dd8427c03733/integrations/tensorflow/iree_tf_compiler/TF/Passes.cpp#L46

File: docs/developers/design_docs/types_and_shapes.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+kernels of dense computation. The phases of compilation can be segmented by
+which MLIR dialects are primarily being transformed:
+
+```
+frontends (PyTorch, JAX, TensorFlow, TOSA, etc.)
+  * Includes user code, serialized ML models / programs, and other libraries
+
+                                    ↓
+
+import dialects (`iree`, `tensor`, `linalg`, etc.)
```
Comment (hanhanW) [inline docs/developers/design_docs/types_and_shapes.md:127]: `IREE` dialect was renamed to `Util` dialect. Maybe delete it or use `standard` or `math` dialect?

File: docs/developers/design_docs/types_and_shapes.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+                                    ↓
+
+`flow` dialect (tensor program modeling and compute workload partitioning)
+
+                                    ↓
+
+`stream` dialect (device placement and asynchronous scheduling)
```
Comment (hanhanW) [inline docs/developers/design_docs/types_and_shapes.md:135]: note that we are handling streaming stuff in flow dialect, which is using `StreamableOpInterface`.

File: docs/developers/design_docs/types_and_shapes.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  types.
+* Supported primitives vary per target backend and may be optionally available.
+  Generally expect for int32 and float32 to be well supported for mobile to
+  desktop -scale devices and for lower or higher bit depth types (e.g. float16,
+  int64) to be optionally available. On embedded systems or certain
+  accelerators there may be no floating support at all.
+
+#### Requirements for host code generation
+
+#### Requirements for device code generation
```
Comment (hanhanW) [inline docs/developers/design_docs/types_and_shapes.md:174]: IIRC, we don't want to generate (big?) memory allocation code on device side either.

File: docs/developers/design_docs/types_and_shapes.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+TODO: LLVM / SPIR-V emulation of types?
```
Comment (hanhanW) [inline docs/developers/design_docs/types_and_shapes.md:176]: Do you aim to describe how codegen backends handle non native types? E.g., if i8 is not available on vulkan target env, how do we emulate i8 load/store?

File: docs/developers/design_docs/types_and_shapes.md (hunk 5 above)
Comment (ScottTodd) [reply docs/developers/design_docs/types_and_shapes.md:127]: Hm, the new architecture diagram still puts "iree" next to "tensor". I can just change it to `standard` here.

File: docs/developers/design_docs/types_and_shapes.md (hunk 6 above)
Comment (ScottTodd) [reply docs/developers/design_docs/types_and_shapes.md:135]: This referring to the upcoming `stream` dialect that @benvanik is working on.

File: docs/developers/design_docs/types_and_shapes.md (hunk 8 above)
Comment (ScottTodd) [reply docs/developers/design_docs/types_and_shapes.md:176]: I have the "Strategies for converting between types" section up top for general information, then my hope was for you / others more familiar with the specifics of each layer to fill in the requirements or strategy preference for that layer down in this section of the doc (e.g. maybe LLVM has efficient support for emulating f64 but SPIR-V does not, so SPIR-V prefers to truncate f64 down to f32).

File: docs/developers/design_docs/types_and_shapes.md (hunk 4 above)
Comment (ScottTodd) [reply docs/developers/design_docs/types_and_shapes.md:34]: I'd prefer for the start of the document to give enough general context: we target different hardware devices and APIs, and some of those have very strict requirements, so the layers can be pretty opinionated.

The HAL section down below can go into specifics for each compiler target / device configuration.

**Comment by benvanik** [comment]
Closing as obsolete. Still would be good to have some documentation about this topic but it's probably worth doing this in the specific frontend documentation instead as frontend dialects have different restrictions themselves.
